test/lib/jdk/test/lib/apps/LingeredApp.java

```diff
@@ -416,6 +416,7 @@ public static void startAppExactJvmOpts(LingeredApp theApp, String... jvmOpts) t
             theApp.runAppExactJvmOpts(jvmOpts);
             theApp.waitAppReady();
         } catch (Exception ex) {
+            theApp.finishApp();
```

Comment:
I'm worried about the output appearing twice in some tests:

```
        try {
            ...
            theApp = new LingeredAppWithLargeArray();
            LingeredApp.startAppExactJvmOpts(theApp, vmArgs);
            attachAndDump(heapDumpFileName, theApp.getPid());
        } finally {
            LingeredApp.stopApp(theApp);
            heapDumpFile.delete();
        }

```
So we always call `stopApp()`:

```
    public void stopApp() throws IOException {
           ...
           if (appProcess != null) {
           ...
            finishApp();
           ...
        }
    }
```


Which means we always call `finishApp()`, which means we always print the output. With your changes won't we see the output twice when there is an exception?

In the CR it looks like `LingeredApp.startAppExactJvmOpts()` is being call from `JCmdTest.java`. I don't see this test. Is it under development? Does it need a `stopApp()` call?

Reply:
It seems error prone to have to call finishApp() manually in order to see the error log. Since LingeredApp.startApp calls finishApp() on exceptions, shouldn't startAppExactJvmOpts do the same thing?

Reply:
Although you have a point, you've also pointed out another problem with this fix.  I think users of `startApp()` are already going to see the output twice because of the `finishApp()` call present there. By adding yet another `finishApp()` call to `startAppExactJvmOpts()`, now they will see it 3 times.

If you want to "move" the `finishApp()` call from `startApp()` to `startAppExactJvmOpts()`, then at least that will maintain the status quo with existing `startApp()` users, but we still have an issue with the output appearing twice, even before this change, and with this change it is now more common as `startAppExactJvmOpts()` will also start seeing it.

Maybe `finishApp()` should maintain an `alreadyCalled` flag so it does nothing on subsequent calls.

Reply:
@plummercj @iklam Thanks for review. I will add a flag for finishApp so we will not call it more than once.

Reply:
I still think the `finishApp()` call should be removed from `startApp()` since it is now in `startAppExactJvmOpts()`. Also move the `println()` to `startAppExactJvmOpts()` and then `startApp()` can just be:

```
             LingeredApp a = new LingeredApp();
             startApp(a, additionalJvmOpts);
```

No need for the exception handling anymore since `startAppExactJvmOpts()` is alread doing that.